### ISSUE TITLE
How to use different weights for forward and backward pass of training

### ISSUE THREAD

**Issue by jwook1004** [tensorpack/tensorpack#916]
Hi,

I was wondering if there's a way to use different weights for the forward and backward pass. 

This question is originated from the DoReFa-Net implementation in Tensorpack, where "remap_variables (func)" is used to quantize weights of the operations (e.g., conv2D, fullyconnected) under the scope using the predefined quantization function, func (as shown in [this](https://github.com/tensorpack/tensorpack/blob/89ce90466bb4186ff8e90cec43b470336a205c26/examples/DoReFa-Net/resnet-dorefa.py#L90)). I checked via [Tensorflow debugger](https://github.com/tensorpack/tensorpack/issues/191) that in the forward and backward convolutions (i.e, Conv2D and Conv2DBackpropInput) the quantized weights are used as the weight for both forward and backward computation. Note that the original (i.e., non-quantized) weight can still be updated with the gradients via [apply_gradients](https://github.com/tensorpack/tensorpack/blob/e943200b1d86cc9254d05118872b40b20c685825/tensorpack/graph_builder/training.py#L174). 

An interesting test-case is to use the quantized weight in the forward pass but use the full-precision weight in the backward pass as motivated by [this paper](https://arxiv.org/abs/1805.11233). This case would be particularly interesting when we try to observe an impact of quantization on the forward and backward passes separately. 

As long as I understood, however, quant-weight-forward/fullprec-weight-backward cannot be implemented via "remap_variables ()", since it replaces the weights with the quantized weights for both forward and backward convolutions. 

Is there any way in Tensorpack to assign different weights for the forward and backward passes of training? Any insights would be very helpful!

Best regards,
Jungwook

**Comment by ppwwyyxx** [comment]
I believe this is a tensorflow question (i.e. how to build such a graph). You'll need to customize the gradients of quantized convolution, just like how the gradient of quantization is customized now in `dorefa.py`. You can refer to `tf.custom_gradient` to learn more.

**Comment by jwook1004** [comment]
I think the case I illustrated is **different** from simply customizing ops in the back-prop.

`tf.custom_gradient` allows change of **computation** in back-prop given the **same** input/output Tensors as arguments. The case I illustrated requires **different** input argument (i.e., full-prec weights)  to be used for the **same** computation in the back-prop.

I was wondering if there's any tensorpack functions / wrapper that can be used/extended to achieve the required change?

**Comment by ppwwyyxx** [comment]
Tensorpack does not and will not provide any non-standard layers like this. Whatever models it supports will be implemented in tensorflow anyway so there is nothing you need from tensorpack.

The thing you described can be implemented with `custom_gradient`. It allows you to do any computation with any tensors in the backward. And in fact it's best to not involve tensorpack layers and implement your customized version of layers in this case, because you'll need to access the full-precision tensors in the gradient function.

**Comment by jwook1004** [comment]
Hi Yuxin,

Thanks for your input. Just to confirm my understanding, I think what you suggested would be similar to the following pseudo code?

@tf.custom_gradient
def convquant(x,w,wq):
  y = conv2d(x,wq)
  def convquant_grad(dy):
    dx = conv2d(dy,w)
    return dx
  return y, convquant_grad

**Comment by ppwwyyxx** [comment]
Yes, something like that should work. These kind of things are usually not very easy to implement correctly so be careful.

**Comment by jwook1004** [comment]
Yeah... I think the above pseudo code still has several problems. For example, it is not clear how to handle weight gradient computation. The above pseudo code does not compute weight gradients (dW), whereas typically instantiation of Conv2D would automatically call Conv2D_backpropInput and Conv2D_backpropFilter. So unless handled separately, the training would not work properly. Also, it would require separate instantiation of variables, which is tricky unless being very careful.

So, it seems to me that Tensorflow might have fundamental limitations with respect to flexibility in manipulating backward-pass computation. I was curious about your thoughts on this point. Would there be a good workaround in Tensorflow to enhance such flexibility?

**Comment by jwook1004** [comment]
Thanks for your input. I agree with you that backward pass of TF can become much flexible with tf.gradients and tf.custom_gradient. 

Regarding one of your comments:
> On this point I think TF is more flexible than any other frameworks. Others only allow you to do (1).

How do you think of Pytorch? I saw that some people also implement quantization schemes on Pytorch by manipulating the back-prop computation (e.g., [link](https://github.com/eladhoffer/quantized.pytorch)). My naive understanding is that one can explicitly define "backward()" function for any op, which is somewhat similar to tf.custom_gradients. I was wondering which one fundamentally allows more flexibility (or maybe they are equally good)?

**Comment by ppwwyyxx** [comment]
They can both do (1). But in tensorflow you can do:
```
@tf.custom_gradient
def func(x, W, QW):
  o1 = f1(x, QW)
  o2 = f2(x, W)
  def grad(dy):
      return tf.gradients(o2, inputs, grad_ys=[dy])
  return o1, grad
```
I don't know if this is easy to do in pytorch or not.

**Comment by ShiRuiUT** [comment]
> Hi,
> 
> I was wondering if there's a way to use different weights for the forward and backward pass.
> 
> This question is originated from the DoReFa-Net implementation in Tensorpack, where "remap_variables (func)" is used to quantize weights of the operations (e.g., conv2D, fullyconnected) under the scope using the predefined quantization function, func (as shown in [this](https://github.com/tensorpack/tensorpack/blob/89ce90466bb4186ff8e90cec43b470336a205c26/examples/DoReFa-Net/resnet-dorefa.py#L90)). I checked via [Tensorflow debugger](https://github.com/tensorpack/tensorpack/issues/191) that in the forward and backward convolutions (i.e, Conv2D and Conv2DBackpropInput) the quantized weights are used as the weight for both forward and backward computation. Note that the original (i.e., non-quantized) weight can still be updated with the gradients via [apply_gradients](https://github.com/tensorpack/tensorpack/blob/e943200b1d86cc9254d05118872b40b20c685825/tensorpack/graph_builder/training.py#L174).
> 
> An interesting test-case is to use the quantized weight in the forward pass but use the full-precision weight in the backward pass as motivated by [this paper](https://arxiv.org/abs/1805.11233). This case would be particularly interesting when we try to observe an impact of quantization on the forward and backward passes separately.
> 
> As long as I understood, however, quant-weight-forward/fullprec-weight-backward cannot be implemented via "remap_variables ()", since it replaces the weights with the quantized weights for both forward and backward convolutions.
> 
> Is there any way in Tensorpack to assign different weights for the forward and backward passes of training? Any insights would be very helpful!
> 
> Best regards,
> Jungwook

Hi, Jungwook. I think this page could help you solve this problem. https://github.com/1202kbs/Understanding-NN/blob/master/models/models_2_4.py
